### DISCUSSION TITLE
Hand-editing the pan in an .h2song file?

### DISCUSSION THREAD

**Discussion by duck57** [hydrogen-music/hydrogen#1123] [Q&A] [answered]
I have a pair of songs I'm working on that both use the circAfriqueV4 drumkit, which has all its samples center-panned.  I'd like to edit the pan information so that all Djembe1 "instruments" have the same pan values (and the same with Djembe 2, etc…) and I figured that editing the file directly would be easier than using the GUI mixer knobs.  The plan is to hand-edit one of the songs until the instruments have a good spatial balance and then copy over the instrument definitions to the other.

What I immediately noticed is that there are separate pan_L and pan_R attributes.  This leaves me with a series of questions that I'll present in descending order of relevance.

1. How do the pan_L and pan_R attributes work?  For instruments that are centered, they are both 1.0.  Should I try to keep their total at 2.0 or would, for example giving pan_L=0.5 and pan_R=0.75 give a right pan and also reduce the total volume?
2. How are invalid values, such as a negative number for pan or random pitch factor, in the .h2song handled?
3. (as a generic rephrase of question 2) Is there documentation on the .h2song format that would list things such as valid value ranges?
4. What ways can someone who does not currently know C++ contribute to Hydrogen?

**Comment by theGreatWhiteShark** [comment, marked as answer]
Hey @duck57,

First of all: what do you find so annoying about the UX that you would consider handling the raw XML files instead of using the GUI? The intended usage would be to tweak the knobs in the mixer until everything feels right, storing the result as a new drumkit, and loading the new drumkit in all your other songs. On the other hand, editing the XML document requires reloading of the song after every edit.

> 1. How do the pan_L and pan_R attributes work? For instruments that are centered, they are both 1.0. Should I try to keep their total at 2.0 or would, for example giving pan_L=0.5 and pan_R=0.75 give a right pan and also reduce the total volume?

It's like a factor of the  volume for the left and the right channel. Placed in the middle (`pan_L` = 0.5 and `pan_R` = 0.5) the song is two times louder than panned all the way to e.g. the right (`pan_L` = 0 and `pan_R` = 0.5). See #665.

There will be some additional pan laws available to choose in the upcoming version of Hydrogen (see #1061) which allow for other volume characteristics while panning.

> 2. How are invalid values, such as a negative number for pan or random pitch factor, in the .h2song handled?

They will be loaded as is from the file and checked when assigned to the `Song` class. If a value is out of range, it should be replaced by its upper/lower bound.

> 3. (as a generic rephrase of question 2) Is there documentation on the .h2song format that would list things such as valid value ranges?

There is a XSD file defining all allowed parameters and ranges for the drumkit part of the `.h2song`.

> 4. What ways can someone who does not currently know C++ contribute to Hydrogen?

The most significant part right now would be the documentation. It's out of date for quite some time and it will need a lot of work to update it. However, it will not need that much in-depth knowledge about the code base but knowing how to handle Hydrogen/what's happening when clicking here and there is usually sufficient. Also, the pull requests will be checked so no harm can be done. I intend to rework some parts too before the release but I can't see to find the time to manage all of it.

In addition, watching this repo and contributing to the discussions (even if its just by giving a thumbs-up in case you like an idea) will help us lots to get a better feeling about how to change/update the UX of Hydrogen.

If you speak a language Hydrogen is not (or improperly) translated to, you can update it.

**Reply by duck57** [reply]
> what do you find so annoying about the UX that you would consider handling the raw XML files instead of using the GUI? The intended usage would be to tweak the knobs in the mixer until everything feels right, storing the result as a new drumkit, and loading the new drumkit in all your other songs. On the other hand, editing the XML document requires reloading of the song after every edit.

The specific drumkit I want to edit has 3 or 4 instruments that represent the same physical drum.  Is there a way for me to select all the instruments for the same drum and adjust all the knobs simultaneously?  The workflow I think I would use is to play a test pattern that uses all the physical drums on loop and then adjust each group of knobs until it sounds good.

The reason I thought of editing the XML directly is that I could ensure that all the instruments corresponding to the same physical drum have equal panning values without having to precisely drag my mouse 3 or 4 times for each adjustment I wish to make.

>  The most significant part right now would be the documentation. It's out of date for quite some time and it will need a lot of work to update it. However, it will not need that much in-depth knowledge about the code base but knowing how to handle Hydrogen/what's happening when clicking here and there is usually sufficient. Also, the pull requests will be checked so no harm can be done. I intend to rework some parts too before the release but I can't see to find the time to manage all of it.

I noticed that a number of screenshots in the manual don't have a one-to-one UI correspondence to what is on my screen.  I assume based on your statement here that it means that the manual's screenshots are probably from old builds of Hydrogen.  Is that accurate?

**Reply by theGreatWhiteShark** [reply]
> The specific drumkit I want to edit has 3 or 4 instruments that represent the same physical drum. Is there a way for me to select all the instruments for the same drum and adjust all the knobs simultaneously? The workflow I think I would use is to play a test pattern that uses all the physical drums on loop and then adjust each group of knobs until it sounds good.
>
> The reason I thought of editing the XML directly is that I could ensure that all the instruments corresponding to the same physical drum have equal panning values without having to precisely drag my mouse 3 or 4 times for each adjustment I wish to make.

I see. No, we don't support setting values of multiple instruments at once.

But we support altering the pan of an instrument strip in the mixer via OSC message.

```bash
oscsend localhost 9000 /Hydrogen/PAN_ABSOLUTE/1 f 0.3
```
sets the pan of the first instrument to 0.3.

You can build a simple shell script propagating the desired pan value to multiple instruments at once.

**Reply by theGreatWhiteShark** [reply]
> I noticed that a number of screenshots in the manual don't have a one-to-one UI correspondence to what is on my screen. I assume based on your statement here that it means that the manual's screenshots are probably from old builds of Hydrogen. Is that accurate?

Yes. This way they serve as hint that the corresponding section most probably needs to be updated too :grin:

**Reply by duck57** [reply]
Thanks for giving me the heads-up about the `oscsend` command.  I haven't been able to make direct progress yet because my system apparently doesn't have oscsend installed:

1.  `oscsend` itself does not seem to be in the Debian repository for armhf 
2.  when I tried to [build oscsend from source](https://fukuchi.org/works/oscsend/index.html.en), the `./configure` step fails every time with `configure: error: cannot guess build type; you must specify one`

However, I found that I can get Hydrogen to set parameters from the knobs from my MIDI keyboard.  I have four knobs on the keyboard, which will work out excellently once `oscsend` is running, as I have four instrument clusters to move and giving each one a knob means I can adjust all of them at once.  I assume that Hydrogen's MIDI learning is strictly one control to one param, so I either need oscsend running or to write some script for a virtual MIDI device that takes each control from my keyboard and multiplexes it into 4 virtual controllers that I map in Hydrogen.

**Reply by theGreatWhiteShark** [reply]
> oscsend itself does not seem to be in the Debian repository for armhf

According to their [packages](https://packages.debian.org/de/buster/armhf/liblo-tools/filelist) webpage it should be available via the `liblo-tools` package.

But you could basically use any other OSC capable software too. For instance some effort was made to make the OSC API of Hydrogen compliant with `TouchOSC` on MacOS.

> However, I found that I can get Hydrogen to set parameters from the knobs from my MIDI keyboard.

Yes. Most controls are supported via both MIDI and OSC. But some, like the session handling commands for quitting Hydrogen, loading or saving a song, are supported via OSC exclusively.

>  I assume that Hydrogen's MIDI learning is strictly one control to one param, so I either need oscsend running or to write some script for a virtual MIDI device that takes each control from my keyboard and multiplexes it into 4 virtual controllers that I map in Hydrogen.

Jup

**Reply by duck57** [reply]
I was successfully able to install `liblo-tools` and write up a script to send OSC messages from my MIDI keyboard and the command line to update multiple instruments at once.  

I'm not sure if it was a problem with my script, something about the way Hydrogen processes OSC messages, or because I ran my test on underpowered hardware, but editing the pan values too quickly had a habit of making Hydrogen lag visually even though my CPU monitor was not at 100%.  Thankfully, the audio playback was steady as ever.

Thanks for your help.

**Reply by theGreatWhiteShark** [reply]
> I'm not sure if it was a problem with my script, something about the way Hydrogen processes OSC messages, or because I ran my test on underpowered hardware, but editing the pan values too quickly had a habit of making Hydrogen lag visually even though my CPU monitor was not at 100%. Thankfully, the audio playback was steady as ever.

Can you open an issue for this and provide the script? I will have a look

**Reply by duck57** [reply]
I finally got around to writing up issue #1217.  The script is rather over-engineered.  Please let me know if you need me to explain it more.

**Comment by oddtime** [comment]
Thanks @theGreatWhiteShark for the accurate answers ;)
Having worked on the new pan implementatio I would add, if that can help @duck57 
that for the instrument pan the correspondence between the _pan knob_ and _panL_ and _panR_ is:
![screenshot-91546f78](https://user-images.githubusercontent.com/54219881/106394860-27c58b00-63ff-11eb-875c-ab34bd81f074.jpg)
where _panL = L_
_panR = R_
pan knob value is expressed with _p_ in [-1,1] from left to right

for note pan it is the same but on the vertical axis the max value is 0.5 instead of 1. (you may check if the max levels are 1 and 0.5)

In Hydrogen 1.0 note and instrument pans act as two BALANCE knobs in a chain, so the meaning of panL panR is really gainL, gainR.

I do find that it is unnecessarily redundant in the next version implementation (where their use will change), but we decided to save pan knob position in the same way
